[wolfssl] update to 5.6.3

JacobBarthelmeh opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/ -->


- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [ ] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.


<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

MonicaLiu0311 commented:
Please get [failure logs here](https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/91381/artifacts?artifactName=failure%20logs%20for%20x64-windows&api-version=7.0&%24format=zip). Please ping me if you need any help.

MonicaLiu0311 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

MonicaLiu0311 commented:
Ping @JacobBarthelmeh 
Is there any new progress?

JacobBarthelmeh commented:
@MonicaLiu0311 Looks like the recent change passed the automated tests. Ready for review. Thanks :)

dg0yt commented on ports/wolfssl/6567.patch:
```diff
@@ -0,0 +1,49 @@
+ # Install the export set
+ install(EXPORT wolfssl-targets
+         DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/wolfssl
+-        FILE wolfssl-targets.cmake
+-        NAMESPACE wolfssl::)
++        FILE wolfssl-config.cmake)
```
Why does this change remove upstream's namespace? We should be glad that they use it.

JacobBarthelmeh replied:
The change was the result of doing a git bisect between the current version and the previous working version. I went through and reverted code from the offending commit until the build of wolfmqtt/wolftpm was successful on Windows with vcpkg.

MonicaLiu0311 replied:
Please keep it as it is here, the upstream has already processed the `wolfssl-config.cmake` file in [CMakeLists.txt#L2354-L2360](https://github.com/wolfSSL/wolfssl/blob/v5.6.3-stable/CMakeLists.txt#L2354-L2360).

JacobBarthelmeh replied:
Ok. I can look at updating wolfmqtt/wolftpm cmake builds. It looks like there is something I can do there to keep in use of namespace with wolfSSL.

dg0yt reviewed (commented):
AFAICT vcpkg prefers raw diffs over patches, i.e. no headers.

MonicaLiu0311 commented:
usage:
```
wolfssl provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(wolfssl CONFIG REQUIRED)
    target_link_libraries(main PRIVATE wolfssl::wolfssl)
```

The following error occurs when testing usage:

```
E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(452,25): error C2065: 'WOLFSSL': undeclared identifier 
E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(452,34): error C2065: 'ssl': undeclared identifier 
E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(452,39): error C2062: type 'char' unexpected 

...

E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(514,40): error C2065: 'WOLFSSL_CTX': undeclared identifier 
E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(514,53): error C2065: 'ctx': undeclared identifier 
E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(514,58): error C2065: 'CallbackIORecv': undeclared identifier 
E:\wolfssl\installed\x64-windows\include\wolfssl\wolfio.h(514,73): error C2146: syntax error: missing ')' before identifier 'CBIORecv' 

...

Rebuild All failed.
```


Fixed by adding the following code in `wolfio.h#50`:
```
#include <wolfssl/ssl.h>
#if defined(USE_WOLFSSL_IO)
    #include <wolfssl/callbacks.h>
#endif
```

JacobBarthelmeh committed:
[wolfssl] update wolfssl to version 5.6.3

[wolfmqtt] update wolfmqtt to version 1.16.0

[wolftpm] add cmake patch for linking to wolfssl version 5.6.3

MonicaLiu0311 commented on ports/wolfmqtt/cmake-patch.diff:
```diff
@@ -0,0 +1,19 @@
++++ b/CMakeLists.txt
+@@ -67,10 +67,13 @@ else()
+         # For support with vcpkg
+         find_package(wolfssl CONFIG)
+         if (wolfssl_FOUND)
++            find_path(WOLFSSL_INCLUDE_DIR NAMES wolfssl/ssl.h)
++            find_library(WOLFSSL_LIBRARY_PATH NAMES wolfssl)
+             target_compile_definitions(wolfmqtt PUBLIC
+                 "ENABLE_MQTT_TLS"
+                 )
+-            target_link_libraries(wolfmqtt PUBLIC wolfssl)
++            include_directories(${WOLFSSL_INCLUDE_DIR})
++            target_link_libraries(wolfmqtt PUBLIC ${WOLFSSL_LIBRARY_PATH})
+         endif()
```
```suggestion
    find_package(wolfssl CONFIG REQUIRED)
    if (wolfssl_FOUND)
        target_link_libraries(main PRIVATE wolfssl::wolfssl)
        target_compile_definitions(wolfmqtt PUBLIC
            "ENABLE_MQTT_TLS"
            )
    endif()
```

It would be better to modify it according to the usage provided by `wolfssl`.

MonicaLiu0311 commented on ports/wolftpm/cmake-patch.diff:
```diff
@@ -0,0 +1,16 @@
++++ b/CMakeLists.txt
+@@ -150,7 +150,10 @@ else()
+         # For support with vcpkg
+         find_package(wolfssl CONFIG)
+         if (wolfssl_FOUND)
+-            target_link_libraries(wolftpm PUBLIC wolfssl)
++            find_path(WOLFSSL_INCLUDE_DIR NAMES wolfssl/ssl.h)
++            find_library(WOLFSSL_LIBRARY_PATH NAMES wolfssl)
++            include_directories(${WOLFSSL_INCLUDE_DIR})
++            target_link_libraries(wolftpm PUBLIC ${WOLFSSL_LIBRARY_PATH})
```
```suggestion
         find_package(wolfssl CONFIG REQUIRED)
         if (wolfssl_FOUND)
             target_link_libraries(wolftpm PUBLIC wolfssl::wolfssl)
```

JacobBarthelmeh replied:
Thanks @MonicaLiu0311 , great suggestions! I updated the patch.

JacobBarthelmeh committed:
[wolfmqtt] Improvements to cmake for linking to wolfSSL
[wolftpm] Improvements to cmake for linking to wolfSSL

MonicaLiu0311 commented:
The usage test passed (header files found):
```
wolfssl provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(wolfssl CONFIG REQUIRED)
    target_link_libraries(main PRIVATE wolfssl::wolfssl)
```
```
wolftpm provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(wolftpm CONFIG REQUIRED)
    target_link_libraries(main PRIVATE wolftpm)
```
```
wolfmqtt provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(wolfmqtt CONFIG REQUIRED)
    target_link_libraries(main PRIVATE wolfmqtt)
```

BillyONeal commented:
> AFAICT vcpkg prefers raw diffs over patches, i.e. no headers.

I don't think this was ever resolved. We were going to make this a rule but @Hoikas convinced us not to. If anything he suggests we should adopt the opposite rule but we don't know how to make the common 'how to author a patch' workflow work in that universe.

BillyONeal commented on ports/wolfmqtt/cmake-patch.diff:
```diff
@@ -0,0 +1,19 @@
+@@ -65,12 +65,12 @@ else()
+         target_compile_options(wolfmqtt PUBLIC ${WOLFSSL_CFLAGS_OTHER})
+     else()
+         # For support with vcpkg
```
Upstream seems to care about us. Is there a bug somewhere about this problem and/or is upstream notified?

BillyONeal replied:
Thanks!

JacobBarthelmeh commented on ports/wolfmqtt/cmake-patch.diff:
```diff
@@ -0,0 +1,19 @@
+                 "ENABLE_MQTT_TLS"
+                 )
+-            target_link_libraries(wolfmqtt PUBLIC wolfssl)
++            target_link_libraries(wolfmqtt PRIVATE wolfssl::wolfssl)
```
Private vs Public is a good question, there is the case where wolfMQTT can be built without TLS and in that case would not use wolfSSL crypto at all. In some of the wolfMQTT examples though we make calls to wolfSSL API. If changing this here to Public helps make it easier to use than I would be okay submitting that change. Though I think an application linking to wolfMQTT could also then link to wolfSSL to pull in any additional wolfSSL API needed?

BillyONeal replied:
If this is just 'supposition' then I think we should leave what upstream did (which was `PUBLIC`) alone; we should respect them unless we have no other means of doing so.

If upstream changed this to PRIVATE that might be fine but I'm nervous about vcpkg changing that in a patch.

BillyONeal replied:
For example, if wolfmqtt installes headers that assume wolfssl headers are present, I think you need PUBLIC here.

BillyONeal replied:
I would be happy if either:

* this changed back to PUBLIC like upstream says, or
* upstream merges https://github.com/wolfSSL/wolfMQTT/pull/345 (which changes to PRIVATE)

BillyONeal reviewed (approved):
If you're happy please mark the PR 'Ready for Review' once more

BillyONeal commented:
Thanks!
